pkg/util/metrics.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"net/http"
+	"strings"
+	"sync"
+
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	prometheus_model "github.com/prometheus/client_model/go"
+	"k8s.io/client-go/util/workqueue"
+)
+
+func CreateValidMetricNameLabel(prefix, name string) string {
+	// "-" aren't valid characters for prometheus metric names or labels
+	return strings.Replace(prefix+name, "-", "_", -1)
+}
+
+// Best effort metric registration with Prometheus.
+func RegisterMetric(metric prometheus.Collector) {
+	if err := prometheus.Register(metric); err != nil {
+		glog.Errorf("Error while registering Prometheus metric: [%v]", err)
+	}
+}
+
+type MetricConfig struct {
+	MetricsEndpoint string
+	MetricsPort     string
+	MetricsPrefix   string
+	MetricsLabels   []string
+}
+
+type PositiveGauge struct {
+	mux         sync.RWMutex
+	name        string
+	gaugeMetric *prometheus.GaugeVec
+}
+
+// Gauge with conditional decrement ensuring its value is never negative.
+func NewPositiveGauge(name string, description string, labels []string) *PositiveGauge {
+	for i, label := range labels {
```

Comment:
Ditto. Let's make and use a copy of `labels`.

Reply:
done